core/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java

```diff
@@ -0,0 +1,255 @@
+                }
+            }
+        }
+        return build;
+    }
+
+    /**
+     * Applies the given settings to all the settings consumers or to none of them. The settings
+     * will be merged with the node settings before they are applied while given settings override existing node
+     * settings.
+     * @param newSettings the settings to apply
+     * @return the unmerged applied settings
+    */
+    public synchronized Settings applySettings(Settings newSettings) {
+        if (lastSettingsApplied != null && newSettings.equals(lastSettingsApplied)) {
+            // nothing changed in the settings, ignore
+            return newSettings;
+        }
+        final Settings build = Settings.builder().put(this.settings).put(newSettings).build();
+        boolean success = false;
+        try {
+            for (SettingUpdater settingUpdater : settingUpdaters) {
+                try {
+                    settingUpdater.prepareApply(build);
+                } catch (Exception ex) {
+                    logger.warn("failed to prepareCommit settings for [{}]", ex, settingUpdater);
+                    throw ex;
+                }
+            }
+            for (SettingUpdater settingUpdater : settingUpdaters) {
+                settingUpdater.apply();
+            }
+            success = true;
+        } catch (Exception ex) {
+            logger.warn("failed to apply settings", ex);
+            throw ex;
+        } finally {
+            if (success == false) {
+                for (SettingUpdater settingUpdater : settingUpdaters) {
+                    try {
+                        settingUpdater.rollback();
+                    } catch (Exception e) {
+                        logger.error("failed to refresh settings for [{}]", e, settingUpdater);
+                    }
+                }
+            }
+        }
+
+        try {
+            for (Map.Entry<String, String> entry : newSettings.getAsMap().entrySet()) {
+                if (entry.getKey().startsWith("logger.")) {
+                    String component = entry.getKey().substring("logger.".length());
+                    if ("_root".equals(component)) {
+                        ESLoggerFactory.getRootLogger().setLevel(entry.getValue());
+                    } else {
+                        ESLoggerFactory.getLogger(component).setLevel(entry.getValue());
+                    }
+                }
+            }
+        } catch (Exception e) {
+            logger.warn("failed to refresh settings for [{}]", e, "logger");
+        }
+
+        return lastSettingsApplied = newSettings;
+    }
+
+    /**
+     * Adds a settings consumer with a predicate that is only evaluated at update time.
+     * <p>
+     * Note: Only settings registered in {@link SettingsModule} can be changed dynamically.
+     * </p>
+     */
+    public synchronized <T> void addSettingsUpdateConsumer(Setting<T> setting, Consumer<T> consumer, Consumer<T> predicate) {
```

Comment:
can we add some java docs explaining what the relation ship of the between the predicate, the consumer and settings application. Also predicate is a generic name, also to me it implies a yes/no decision where it seems that we expect it to throw an exception if it doesn't like things. Maybe call it a validator and doc how it is expected to work?